external/fv3fit/fv3fit/_shared/models.py

```diff
@@ -238,3 +240,76 @@ def load(cls, path: str) -> "TransformedPredictor":
         ]
         transformed_model = cls(base_model, transform_configs)
         return transformed_model
+
+
+@io.register("out_of_sample")
+class OutOfSampleModel(Predictor):
+
+    _CONFIG_FILENAME = "out_of_sample_model.yaml"
+    _BASE_MODEL_SUBDIR = "base"
+    _NOVELTY_DETECTOR_SUBDIR = "novelty"
+
+    def __init__(
+        self,
+        base_model: Predictor,
+        novelty_detector: NoveltyDetector,
+        cutoff: float = None,
+    ):
+        """
+        Args:
+            base_model: trained ML-based predictor, to be used unless an input is
+                deemed "out of sample"
+            novelty_detection: trained novelty detector on the same inputs, which
+                predicts whether a training sample does _not_ belong the training
+                distribution of base_model
+        """
+        self.base_model = base_model
+        self.novelty_detector = novelty_detector
+        self.cutoff = cutoff
+
+        output_variables = base_model.output_variables
+        base_inputs = set(base_model.input_variables)
+        novelty_inputs = set(novelty_detector.input_variables)
+        input_variables = tuple(sorted(base_inputs.union(novelty_inputs)))
+        super().__init__(
+            input_variables=input_variables, output_variables=output_variables
+        )
+
+    def predict(self, X: xr.Dataset) -> xr.Dataset:
+        base_predict = self.base_model.predict(X)
+        novelties = self.novelty_detector.predict_novelties(X, self.cutoff)
+        masked_predict = base_predict.copy()
+        for output_variable in self.output_variables:
+            masked_predict[output_variable] = xr.where(
+                novelties[NoveltyDetector._NOVELTY_OUTPUT_VAR] == 0,
+                base_predict[output_variable],
+                0,
+            )
+        return masked_predict
+
+    def dump(self, path: str):
```

Comment:
For models that have to be manually constructed like this and the `EnsembleModel` I can't think of a use case where the dump method is needed- unless there's one you can think of, I'd suggest raising a `NotImplementedError` that points to the documentation for how to write the manual configuration, e.g. https://github.com/ai2cm/fv3net/blob/9b2b61d304d6494c3889578a211ed9da5d5d89b1/external/fv3fit/fv3fit/_shared/models.py#L168